### PR TITLE
Add feature level security

### PR DESCRIPTION
**Pull request by stevenpears-bjss**
As well as (properly) implementing fine-grained feature level security, this PR remediates the following 3 findings from the October 2022 IT Health Check (ITHC):

- [CB2-6385](https://dvsa.atlassian.net/browse/CB2-6385) - MOTH and Search MC roles are overprivileged
- [CB2-6412](https://dvsa.atlassian.net/browse/CB2-6412) - JWT's are not correctly verified
- [CB2-6470](https://dvsa.atlassian.net/browse/CB2-6470) - Assigned permissions aren't checked correctly

### CODE REVIEW

File: src/functions/rolePolicyFactory.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import Role, { getLegacyRoles } from "../services/roles";
+import StatementBuilder from "../services/StatementBuilder";
+import newPolicyDocument from "./newPolicyDocument";
+
+type NonEmptyArray<T> = [T, ...T[]];
+
+const accessToHttpVerbs: AccessHttpVerbMap = {
+  read: ["GET", "HEAD"],
+  write: ["*"],
+  view: ["GET"],
```
Comment (aaron-downing) [inline src/functions/rolePolicyFactory.ts:14]: If this is the legacy factory then we should be able to remove `view: ["GET"]`, the legacy roles strictly used `read` and `write`

File: src/functions/functionalPolicyFactory.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      path: "vehicles/*",
+    },
+  ],
+};
+
+function toStatement(access: IApiAccess): Statement {
+  return new StatementBuilder().setEffect("Allow").setHttpVerb(access.verb).setResource(access.path).build();
+}
+
+export function generatePolicy(jwt: any, logEvent: ILogEvent): APIGatewayAuthorizerResult | undefined {
```
Comment (aaron-downing) [inline src/functions/functionalPolicyFactory.ts:37]: Would be nice to have some unit tests for this new function

File: tests/unit/services/roles.unitTest.ts [hunk 3]
```diff
@@ -16,7 +16,7 @@ describe("getValidRoles()", () => {
   });
 
   it("should return list of valid roles for TechRecord.View", () => {
-    const roles: Role[] = getValidRoles(tokenWithRoles(["TechRecord.View"]));
+    const roles: Role[] = getLegacyRoles(tokenWithRoles(["TechRecord.View"]), {});
```
Comment (aaron-downing) [inline tests/unit/services/roles.unitTest.ts:19]: This test should likely be removed as it's not legacy

File: src/functions/functionalPolicyFactory.ts (hunk 2 above)
Comment (aaron-downing) [reply src/functions/functionalPolicyFactory.ts:37]: Realised you've somewhat tested this function through the new `authoriser` tests but would be good to see that the ARN policies look as expected and not just checking that the length is equal to 3 in this test
`it("should return multiple statements when multiple roles are valid", async () => {`

Comment (stevenpears-bjss) [reply src/functions/functionalPolicyFactory.ts:37]: Added test to ensure policy format is being produced correctly

File: src/functions/authorizer.ts [hunk 4]
```diff
@@ -29,69 +17,27 @@ const accessToHttpVerbs: AccessHttpVerbMap = {
...
-      writeLogMessage(logEvent, JWT_MESSAGE.INVALID_ROLES);
-      return unauthorisedPolicy();
-    }
-    // by this point we know authorizationToken meets formatting requirements
-    // remove 'Bearer ' when verifying signature
-    await checkSignature(event.authorizationToken.substring(7), jwt);
+    const jwt: any = await getValidJwt(event.authorizationToken, logEvent);
 
-    let statements: Statement[] = [];
+    const policy = generateRolePolicy(jwt, logEvent) ?? (await generateFunctionalPolicy(jwt, logEvent));
```
Comment (andrewf-bjss) [inline src/functions/authorizer.ts:26]: Nitpick: this either/or `??` could be extracted to a `generatePolicy()`...  the arguments and return type are the same so it doesn't matter to the caller if they get a Role or Functional policy back (though putting a Factory in front of a Factory might be more confusing!)

File: src/functions/functionalPolicyFactory.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      path: "vehicles/*",
+    },
+  ],
+  "TechRecord.View": [
+    {
+      verb: "GET",
+      path: "vehicles/*",
+    },
+  ],
+};
```
Comment (andrewf-bjss) [inline src/functions/functionalPolicyFactory.ts:31]: 2 Things:
1. Is this the best place to maintain this list of VTM specific functional roles? (a separate config only file might be easier to discover perhaps)

2. There are some `TestResult.*` permissions that we might wish to include in this initial implementation:
```suggestion
const functionConfig: { [key: string]: NonEmptyArray<IApiAccess> } = {
  "TechRecord.Amend": [
    {
      verb: "POST",
      path: "vehicles/*",
    },
    {
      verb: "PUT",
      path: "vehicles/*",
    },
  ],
  "TechRecord.View": [
    {
      verb: "GET",
      path: "vehicles/*",
    },
  ],
  "TestResult.Amend": [
    {
      verb: "POST",
      path: "test-result/*",
    },
    {
      verb: "PUT",
      path: "test-result/*",
    },
    {
      verb: "GET",
      path: "test-types/*",
    },
    {
      verb: "GET",
      path: "test-stations/*",
    },
    {
      verb: "GET",
      path: "defects/*",
    },
    {
      verb: "GET",
      path: "reference/*",
    },
  ],
  "TestResult.View": [
    {
      verb: "GET",
      path: "test-result/*",
    },
    {
      verb: "GET",
      path: "test-types/*",
    },
    {
      verb: "GET",
      path: "v1/document-retrieval/*",
    },
};
```

File: src/functions/authorizer.ts (hunk 4 above)
Comment (stevenpears-bjss) [reply src/functions/authorizer.ts:26]: At first glance I'm not sure what benefit that gives? This is a one liner to show priority...could you help me see how abstracting this away helps?

File: src/functions/functionalPolicyFactory.ts (hunk 5 above)
Comment (stevenpears-bjss) [reply src/functions/functionalPolicyFactory.ts:31]: knew this wasn't the full list so I really appreciate you helping me out with that! 👍 
Can't see a downside to exporting the config into its own file - I'll get that done next.

File: src/services/tokens.ts [hunk 6]
```diff
@@ -1,31 +1,39 @@
...
+  await checkSignature(authorizationToken, decoded, tenantId, clientId);
+
   let username;
+  const payload = decoded.payload as CVSJWTPayload;
 
-  if (decoded.payload.unique_name) {
-    username = decoded.payload.unique_name;
-  } else if (decoded.payload.preferred_username) {
-    username = decoded.payload.preferred_username;
+  if (decoded.payload) {
```
Comment (tom-bjss) [inline src/services/tokens.ts:27]: nit: can use variable on line 25

File: src/services/signature-check.ts [hunk 7]
```diff
@@ -1,13 +1,13 @@
 import * as JWT from "jsonwebtoken";
 import { getCertificateChain } from "./azure";
 
-export const checkSignature = async (encodedToken: string, decodedToken: any): Promise<void> => {
+export const checkSignature = async (encodedToken: string, decodedToken: any, tenantId: string, clientId: string): Promise<void> => {
```
Comment (tom-bjss) [inline src/services/signature-check.ts:4]: decodedtoken is an any here and can be a JWT as seen elsewhere

File: src/services/roles.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import { logEvent } from "../functions/authorizer";
+import { ILogEvent } from "../models/ILogEvent";
 
 export type Access = "read" | "write" | "view";
```
Comment (tom-bjss) [inline src/services/roles.ts:4]: Im sure this function can just be improved but looks out of scope of the changes youve made.

File: src/services/tokens.ts [hunk 9]
```diff
@@ -1,31 +1,39 @@
...
+import { ILogEvent } from "../models/ILogEvent";
+import { checkSignature } from "./signature-check";
 
-export const getValidJwt = (authorizationToken: string): any => {
+interface CVSJWTPayload extends JwtPayload {
+  unique_name: string;
+  preferred_username: string;
+}
+
+export const getValidJwt = async (authorizationToken: string, logEvent: ILogEvent, tenantId: string, clientId: string): Promise<any> => {
```
Comment (tom-bjss) [inline src/services/tokens.ts:11]: can this function now return a Promise<JWT> as it returns decoded which now has a type. this would then bubble up to the methods that use this function

File: src/functions/authorizer.ts [hunk 10]
```diff
@@ -29,69 +17,32 @@ const accessToHttpVerbs: AccessHttpVerbMap = {
...
-      reportNoValidRoles(jwt, event, context);
-      writeLogMessage(logEvent, JWT_MESSAGE.INVALID_ROLES);
-      return unauthorisedPolicy();
-    }
-    // by this point we know authorizationToken meets formatting requirements
-    // remove 'Bearer ' when verifying signature
-    await checkSignature(event.authorizationToken.substring(7), jwt);
+  try {
+    initialiseLogEvent(event);
+    const jwt: any = await getValidJwt(event.authorizationToken, logEvent, process.env.AZURE_TENANT_ID, process.env.AZURE_CLIENT_ID);
```
Comment (tom-bjss) [inline src/functions/authorizer.ts:29]: see comment on getValidJwt function which would return a type now

File: src/services/tokens.ts (hunk 6 above)
Comment (stevenpears-bjss) [reply src/services/tokens.ts]: Had to alter the structure a little, but that's resolved now

File: src/functions/authorizer.ts [hunk 11]
```diff
@@ -29,69 +17,32 @@ const accessToHttpVerbs: AccessHttpVerbMap = {
...
+    if (policy !== undefined) {
+      return policy;
     }
 
-    writeLogMessage(logEvent);
-    return {
-      principalId: jwt.payload.sub,
-      policyDocument: newPolicyDocument(statements),
-    };
+    reportNoValidRoles(jwt, event, context, logEvent);
```
Comment (tom-bjss) [inline src/functions/authorizer.ts:37]: as above

File: tests/unit/functions/authoriser.unitTest.ts [hunk 12]
```diff
@@ -1,15 +1,11 @@
-import { APIGatewayTokenAuthorizerEvent, Context } from "aws-lambda";
+import { APIGatewayTokenAuthorizerEvent, BaseStatement, Context, MaybeStatementResource, Statement, StatementAction } from "aws-lambda";
```
Comment (tom-bjss) [inline tests/unit/functions/authoriser.unitTest.ts:1]: Solid refactor in this file Steve

**Review by tom-bjss — COMMENTED** [review]
Good work, just some comments about type safety now that a core function return a correct type instead of any

File: src/services/tokens.ts (hunk 9 above)
Comment (stevenpears-bjss) [reply src/services/tokens.ts]: done

**Comment by stevenpears-bjss** [comment]
All `any` types removed from code - with the exception of handling errors in the catch

**Review by tom-bjss — APPROVED** [review]

File: src/functions/rolePolicyFactory.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+const Configuration: AuthorizerConfig = {
+  CVSFullAccess: ["/*"],
+  CVSPsvTester: ["/*"],
+  CVSHgvTester: ["/*"],
+  CVSAdrTester: ["/*"],
+  CVSTirTester: ["/*"],
+  VTMAdmin: ["/*"],
+  Certs: ["/*/v1/document-retrieval", "/*/v1/document-retrieval/*"],
+  VehicleData: ["/*/v1/enquiry", "/*/v1/enquiry/*"],
+  DVLATrailers: ["/*/trailers", "/*/trailers/*"],
```
Comment (aaron-downing) [inline src/functions/rolePolicyFactory.ts:26]: Shouldn't these endpoints just be:
```
  Certs: ["/v1/document-retrieval", "/v1/document-retrieval/*"],
  VehicleData: ["/v1/enquiry", "/v1/enquiry/*"],
  DVLATrailers: ["/v1/trailers", "/v1/trailers/*"],
```
without the `/*` prefix with the way the APIG routes are setup currently
image

File: src/functions/authorizer.ts (hunk 4 above)
Comment (andrewf-bjss) [reply src/functions/authorizer.ts:33]: it's a nitpick, so I'm happy with the code as-is - resolving.

**Review by andrewf-bjss — APPROVED** [review]

File: src/functions/rolePolicyFactory.ts (hunk 13 above)
Comment (Ian-Hodges) [inline src/functions/rolePolicyFactory.ts:26]: Do Certs, VehicleData, and DVLATrailers need adding to backwardsCompatibleRoleNames?
image

Comment (stevenpears-bjss) [reply src/functions/rolePolicyFactory.ts:26]: These aren't URL paths, they're ARN paths (there's several more fragments to an ARN path including API stage and method).
The reason for the prefix `*` was to get around the string splitting already in place - it didn't correctly split the string without it.
That said there should be new tests added to confirm that - I'll add them in 👍

Comment (stevenpears-bjss) [reply src/functions/rolePolicyFactory.ts:26]: That's the existing codebase, so they're not in there today...I guess I add a test and find out?

Comment (stevenpears-bjss) [reply src/functions/rolePolicyFactory.ts]: @aaron-downing absolutely right - I got the number of `*`'s wrong and added one at the beginning incorrectly. Now fixed 👍

Comment (stevenpears-bjss) [reply src/functions/rolePolicyFactory.ts]: @Ian-Hodges it appears these three roles can't have previously worked as they weren't in the list. Now added them 👍

**Review by aaron-downing — APPROVED** [review]

File: src/functions/rolePolicyFactory.ts (hunk 13 above)
Comment (Ian-Hodges) [reply src/functions/rolePolicyFactory.ts]: It looks like there are no unit tests for the legacy roles. You could do the lot with the following.
```
it.each([
    ['CVSFullAccess', '/*', 1],
    ['CVSPsvTester', '/*', 1],
    ['CVSHgvTester', '/*', 1],
    ['CVSAdrTester', '/*', 1],
    ['CVSTirTester', '/*', 1],
    ['VTMAdmin', '/*', 1],
    ['Certs', '/v1/document-retrieval/*', 2],
    ['VehicleData', '/v1/enquiry/*', 2],
    ['DVLATrailers', '/v1/trailers/*', 2]])(
    "legacy roles should return valid write statements on valid JWT", async (legacyRolename, arn, statementCount) => {
      const jwtJsonClone = JSON.parse(JSON.stringify(jwtJson));
      jwtJsonClone.payload.roles = [legacyRolename];
      (getValidJwt as jest.Mock) = jest.fn().mockReturnValue(jwtJsonClone);

      const returnValue: APIGatewayAuthorizerResult = await authorizer(event, exampleContext());

      expect(returnValue.principalId).toEqual(jwtJson.payload.sub);

      expect(returnValue.policyDocument.Statement.length).toEqual(statementCount);
      expect(returnValue.policyDocument.Statement).toContainEqual({
        Effect: "Allow",
        Action: "execute-api:Invoke",
        Resource: `arn:aws:execute-api:eu-west-1:*:*/*/*${arn}`,
      });
    });
```

Comment (stevenpears-bjss) [reply src/functions/rolePolicyFactory.ts]: The authoriser tests all rely on correctly processing legacy roles

Comment (Ian-Hodges) [reply src/functions/rolePolicyFactory.ts]: Are you sure? The code that uses backwardsCompatibleRoleNames is not covered?
image

Comment (stevenpears-bjss) [reply src/functions/rolePolicyFactory.ts]: @Ian-Hodges I just added a test that uses `DVLATrailers`, I had to specifically manipulate the array you're talking about to get it to work - so I know it's used in the test I wrote.

The test fakes the JWT to make sure the roles are added, but otherwise performs an end to end call of the authoriser.

As to whether or not every logic path is covered, I'm sure there are gaps as this is the legacy codebase and not the purpose of this PR.

I'm happy to help you work on increasing the overall test coverage if that's something you'd like to help with - but I do need to focus on the tickets we're working on right now 👍

Comment (Ian-Hodges) [reply src/functions/rolePolicyFactory.ts]: To confirm, are the legacy roles in the jwt CVSFullAccess, CVSPsvTester etc... or are they CVSFullAccess.read, CVSFullAccess.write, CVSPsvTester.read, CVSPsvTester.write etc...

**Review by Ian-Hodges — APPROVED** [review]

File: src/functions/rolePolicyFactory.ts (hunk 13 above)
Comment (Ian-Hodges) [reply src/functions/rolePolicyFactory.ts]: I think I can see where I've misunderstood things. I thought the backwardsCompatibleRoleNames logic was part of your work to support legacy roles. It is in fact a previous developer doing the same thing when CVS switched roles from a `name ` format to a `name.access` format. I suspect it's redundant now and can be removed, but that's for another day.
